File: examples/clojure/02-variables-and-types/main.clj
;;; Variables and Types in Clojure
;;; ==============================
;;; Clojure uses dynamic, strong typing.
;;; All core data structures are immutable and persistent.
;;; Clojure is hosted on the JVM and interoperates with Java types.

(ns main)

;; ── Basic types ──────────────────────────────────────

(println "--- Basic types ---")

;; Integers — uses Java Long (64-bit), auto-promotes to BigInteger
(def count' 42)
(def big-num 9000000000000000000N)  ; BigInteger literal with N
(println "count =" count' "type =" (type count'))
(println "big-num =" big-num "type =" (type big-num))

;; Floats — Java Double
(def pi' 3.14159)
(println "pi =" pi' "type =" (type pi'))

;; Rationals — exact arithmetic!
(def third (/ 1 3))
(println "1/3 =" third "type =" (type third))
(println "1/3 + 1/3 + 1/3 =" (+ third third third))  ; exactly 1

;; Booleans
(def active true)
(println "active =" active "type =" (type active))

;; Strings — Java String
(def name' "Software Atlas")
(println "name =" name' "type =" (type name'))

;; Characters
(def letter \A)
(println "letter =" letter "type =" (type letter))

;; nil — absence of a value
(def nothing nil)
(println "nothing =" nothing "nil? =" (nil? nothing))

;; Keywords — lightweight identifiers (like atoms in Erlang)
(def status :active)
(println "status =" status "type =" (type status))

;; ── Immutability ─────────────────────────────────────

(println "\n--- Immutability ---")

;; def creates a global binding — not a mutable variable
(def x 42)
(println "x =" x)
;; (def x 43) would rebind the var, not mutate — and is discouraged

;; let creates local, immutable bindings
(let [a 10
      b (+ a 5)]
  (println "a =" a "b =" b))
;; a and b don't exist outside this let block

;; ── Strong typing ────────────────────────────────────

(println "\n--- Strong typing ---")

;; Clojure does NOT implicitly convert between types
;; (+ 1 "2") → ClassCastException!
(try
  (+ 1 "2")
  (catch Exception e
    (println "(+ 1 \"2\") →" (.getSimpleName (class e)))))

;; Explicit conversions
(println "(Integer/parseInt \"42\") =" (Integer/parseInt "42"))
(println "(str 42) =" (str 42))
(println "(double 42) =" (double 42))

;; ── Vectors ──────────────────────────────────────────

(println "\n--- Vectors ---")

;; Vectors — indexed, immutable sequences
(def nums [1 2 3 4 5])
(println "nums =" nums)
(println "count =" (count nums))
(println "(first nums) =" (first nums))
(println "(last nums) =" (last nums))
(println "(nth nums 2) =" (nth nums 2))

;; "Adding" returns a NEW vector (structural sharing)
(def extended (conj nums 6))
(println "(conj nums 6) =" extended)
(println "nums still =" nums "← unchanged!")

;; ── Lists ────────────────────────────────────────────

(println "\n--- Lists ---")

;; Lists — linked lists, efficient prepend
(def fruits '("apple" "banana" "cherry"))
(println "fruits =" fruits)
(println "(first fruits) =" (first fruits))
(println "(rest fruits) =" (rest fruits))

;; conj on lists prepends (not appends like vectors)
(println "(conj fruits \"date\") =" (conj fruits "date"))

;; ── Maps ─────────────────────────────────────────────

(println "\n--- Maps ---")

;; Maps — key-value pairs, keywords as keys
(def person {:name "Ada" :age 36 :active true})
(println "person =" person)

;; Access — keywords are functions!
(println "(:name person) =" (:name person))
(println "(:age person) =" (:age person))

;; Missing key returns nil (or a default)
(println "(:email person) =" (:email person))          ; nil
(println "(:email person \"?\") =" (:email person "?")) ; default

;; "Updating" returns a NEW map
(def updated (assoc person :age 37))
(println "(assoc person :age 37) =" updated)
(println "person still =" person "← unchanged!")

;; Nested update
(def nested {:user {:name "Ada" :prefs {:theme "dark"}}})
(def changed (assoc-in nested [:user :prefs :theme] "light"))
(println "assoc-in =" changed)

;; ── Sets ─────────────────────────────────────────────

(println "\n--- Sets ---")

(def tags #{:clojure :fp :jvm})
(println "tags =" tags)
(println "(contains? tags :fp) =" (contains? tags :fp))
(println "(conj tags :lisp) =" (conj tags :lisp))

;; ── Destructuring ────────────────────────────────────

(println "\n--- Destructuring ---")

;; Vector destructuring
(let [[a b c] [10 20 30]]
  (println "a =" a "b =" b "c =" c))

;; Map destructuring
(let [{:keys [name age]} {:name "Ada" :age 36}]
  (println "name =" name "age =" age))

;; With defaults
(let [{:keys [name email] :or {email "n/a"}} {:name "Ada"}]
  (println "name =" name "email =" email))

;; ── Sequences (the universal abstraction) ────────────

(println "\n--- Sequences ---")

;; Almost everything is a sequence in Clojure
;; map, filter, reduce work on vectors, lists, maps, sets, strings...

(println "(map inc [1 2 3]) =" (map inc [1 2 3]))
(println "(filter even? (range 10)) =" (filter even? (range 10)))
(println "(reduce + [1 2 3 4 5]) =" (reduce + [1 2 3 4 5]))

;; Lazy sequences
(println "(take 5 (range)) =" (take 5 (range)))  ; infinite sequence!

;; ── Java interop ─────────────────────────────────────

(println "\n--- Java interop ---")

;; Clojure values are Java objects
(println "(type 42) =" (type 42))           ; java.lang.Long
(println "(type \"hi\") =" (type "hi"))     ; java.lang.String
(println "(type true) =" (type true))       ; java.lang.Boolean
(println "(type :key) =" (type :key))       ; clojure.lang.Keyword
(println "(type [1 2]) =" (type [1 2]))     ; clojure.lang.PersistentVector

;; Call Java methods
(println "(.toUpperCase \"hello\") =" (.toUpperCase "hello"))
(println "(Math/sqrt 2) =" (Math/sqrt 2))

;; ── Summary ──────────────────────────────────────────

(println "\n--- Summary ---")
(println "Clojure: dynamic + strong typing")
(println "  - Types checked at runtime (dynamic)")
(println "  - No implicit conversions (strong)")
(println "  - All data structures immutable by default")
(println "  - Persistent data structures (structural sharing)")
(println "  - Keywords as lightweight identifiers")
(println "  - Maps as primary data structure")
(println "  - Sequences as universal abstraction")
(println "  - Hosted on JVM — Java interop")
(println "  - Exact arithmetic (rationals, BigInteger)")

File: examples/erlang/02-variables-and-types/main.erl
%%% Variables and Types in Erlang
%%% ==============================
%%% Erlang uses dynamic, strong typing.
%%% Variables start with Uppercase and are single-assignment (immutable).

-module(main).
-export([start/0]).

start() ->
    %% ── Basic types ──────────────────────────────────

    %% Integers — arbitrary precision (no overflow)
    Count = 42,
    Big = math:pow(10, 100),  % returns float for large numbers
    io:format("Count = ~p~n", [Count]),
    io:format("Big = ~p~n", [Big]),

    %% Floats — IEEE 754 double precision
    Pi = 3.14159,
    io:format("Pi = ~p~n", [Pi]),

    %% Atoms — symbolic constants (like symbols in Ruby/Lisp)
    Status = ok,
    Color = red,
    io:format("Status = ~p, Color = ~p~n", [Status, Color]),

    %% Booleans — atoms 'true' and 'false'
    Active = true,
    io:format("Active = ~p~n", [Active]),

    %% Strings — actually lists of integers (Unicode code points)
    Name = "Software Atlas",
    io:format("Name = ~p~n", [Name]),
    io:format("Name as list = ~w~n", [Name]),  % shows integer list

    %% ── Single assignment (immutability) ─────────────

    io:format("~n--- Single assignment ---~n"),

    X = 42,
    %% X = 43,  % ERROR: no match of right hand side value 43
    %% Once X is bound to 42, it cannot be rebound!

    io:format("X = ~p (cannot change!)~n", [X]),

    %% But you can use different variable names
    Y = X + 1,
    io:format("Y = X + 1 = ~p~n", [Y]),

    %% ── Pattern matching ─────────────────────────────

    io:format("~n--- Pattern matching ---~n"),

    %% = is the MATCH operator, not assignment!
    %% It binds unbound variables AND checks equality for bound ones

    {ok, Value} = {ok, 42},
    io:format("{ok, Value} = {ok, 42} → Value = ~p~n", [Value]),

    %% Pattern matching on structure
    {point, PX, PY} = {point, 3.0, 4.0},
    io:format("Point = (~p, ~p)~n", [PX, PY]),

    %% Match failure raises an error
    %% {ok, _} = {error, "oops"},  % ERROR: no match

    %% ── Tuples ───────────────────────────────────────

    io:format("~n--- Tuples ---~n"),

    %% Tuples are fixed-size, heterogeneous collections
    %% Convention: first element is an atom tag
    Point = {point, 3.0, 4.0},
    Person = {person, "Ada", 36},
    Result = {ok, "success"},
    Error = {error, not_found},

    io:format("Point = ~p~n", [Point]),
    io:format("Person = ~p~n", [Person]),
    io:format("Result = ~p~n", [Result]),
    io:format("Error = ~p~n", [Error]),

    %% Tuple size and element access
    io:format("tuple_size(Point) = ~p~n", [tuple_size(Point)]),
    io:format("element(2, Point) = ~p~n", [element(2, Point)]),

    %% ── Lists ────────────────────────────────────────

    io:format("~n--- Lists ---~n"),

    %% Lists are variable-length linked lists
    Nums = [1, 2, 3, 4, 5],
    Fruits = ["apple", "banana", "cherry"],

    io:format("Nums = ~p~n", [Nums]),
    io:format("Fruits = ~p~n", [Fruits]),
    io:format("length(Nums) = ~p~n", [length(Nums)]),

    %% Head and tail
    [H | T] = Nums,
    io:format("[H | T] = ~p → H = ~p, T = ~p~n", [Nums, H, T]),

    %% Prepend (O(1))
    Extended = [0 | Nums],
    io:format("[0 | Nums] = ~p~n", [Extended]),

    %% List concatenation
    Joined = [1, 2] ++ [3, 4],
    io:format("[1,2] ++ [3,4] = ~p~n", [Joined]),

    %% List comprehension
    Evens = [N || N <- lists:seq(1, 20), N rem 2 == 0],
    io:format("Evens 1..20 = ~p~n", [Evens]),

    %% ── Maps (Erlang 17+) ────────────────────────────

    io:format("~n--- Maps ---~n"),

    Ages = #{ada => 36, grace => 85},
    io:format("Ages = ~p~n", [Ages]),

    %% Access
    AdaAge = maps:get(ada, Ages),
    io:format("Ages[ada] = ~p~n", [AdaAge]),

    %% Missing key → error (use maps:get/3 for default)
    Default = maps:get(alan, Ages, 0),
    io:format("Ages[alan] (default 0) = ~p~n", [Default]),

    %% Update (returns new map — immutable!)
    Updated = Ages#{ada := 37},
    io:format("Updated = ~p~n", [Updated]),
    io:format("Original = ~p (unchanged!)~n", [Ages]),

    %% ── Strong typing (no implicit conversions) ──────

    io:format("~n--- Strong typing ---~n"),

    %% Erlang does NOT implicitly convert between types
    %% 1 + "2" → ERROR: bad argument in an arithmetic expression

    %% Explicit conversions
    IntVal = 42,
    StrVal = integer_to_list(IntVal),
    io:format("integer_to_list(42) = ~p~n", [StrVal]),

    FloatVal = float(IntVal),
    io:format("float(42) = ~p~n", [FloatVal]),

    %% ── Type checking with guards ────────────────────

    io:format("~n--- Type checking (guards) ---~n"),

    describe(42),
    describe(3.14),
    describe(hello),
    describe("text"),
    describe({point, 1, 2}),

    %% ── No null ──────────────────────────────────────

    io:format("~n--- No null ---~n"),

    %% Erlang has no null. Use atoms or tagged tuples:
    Found = {ok, 42},
    Missing = {error, not_found},
    Undefined = undefined,

    io:format("Found = ~p~n", [Found]),
    io:format("Missing = ~p~n", [Missing]),
    io:format("Undefined = ~p~n", [Undefined]),

    %% Pattern match to handle each case
    handle_result(Found),
    handle_result(Missing),

    %% ── Summary ──────────────────────────────────────

    io:format("~n--- Summary ---~n"),
    io:format("Erlang: dynamic + strong typing~n"),
    io:format("  - Types checked at runtime (dynamic)~n"),
    io:format("  - No implicit conversions (strong)~n"),
    io:format("  - Single assignment (immutable variables)~n"),
    io:format("  - Pattern matching instead of assignment~n"),
    io:format("  - Atoms for symbolic constants~n"),
    io:format("  - Tagged tuples: {ok, V}, {error, R}~n"),
    io:format("  - Arbitrary-precision integers~n"),
    io:format("  - No null~n"),
    ok.

%% ── Helper functions ─────────────────────────────────

describe(V) when is_integer(V) ->
    io:format("  ~p is an integer~n", [V]);
describe(V) when is_float(V) ->
    io:format("  ~p is a float~n", [V]);
describe(V) when is_atom(V) ->
    io:format("  ~p is an atom~n", [V]);
describe(V) when is_list(V) ->
    io:format("  ~p is a list~n", [V]);
describe(V) when is_tuple(V) ->
    io:format("  ~p is a tuple~n", [V]);
describe(V) ->
    io:format("  ~p is something else~n", [V]).

handle_result({ok, V}) ->
    io:format("  Result: got ~p~n", [V]);
handle_result({error, Reason}) ->
    io:format("  Result: error ~p~n", [Reason]).
